File: boscof-cs50-2017-x-credit-20170411T153408Z/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long long n;
    //ask user for valid input
    do
    {
        printf("Number:\n");
        n = get_long_long();
    }
    while (n<0);
        
    long long m=n;
    int product=0;
    int even_sum=0;
    //multiply every second digit by two, starting from the number's second-to-last digit
    do
    {
        m/=10;
        product = (m % 10)*2;
        //sums the products digits
        do
        {
            even_sum = even_sum + product % 10;
            product/=10;
        }
        while(product>0);
        m/=10;
    }
    while (m>0);
    
    m=n;
    int odd_sum=0;
    //sum every second digit, starting from the numbers last digit
    do
    {
        odd_sum = odd_sum + (m % 10);
        m/=100;
    }
    while (m>0);
    int sum=odd_sum + even_sum;
    
    int count=0;
    m=n;
    //counts the number of digits in the original number
    while (m!=0)
    {
        m/=10;
        ++count;
    }
    m=n;
    
    m=n;
    //assigns the first two digits of the number to new variables
    while(m>100)
    {
        m/=10;
    }
    int two_digit=m;
    int first_digit=two_digit/10;
    
    //checks the validity of the credit card number
    if (sum % 10==0)
    {
        //determines the credit card type
        if ((count==13 && first_digit==4)||(count==16 && first_digit==4))
        {
            printf("VISA\n");
        }
        else if (count==16 && (two_digit==51||two_digit==52||two_digit==53||two_digit==54||two_digit==55))
        {
            printf("MASTERCARD\n");
        }
        else if (count==15 && (two_digit==34||two_digit==37))
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: boscof-cs50-2017-x-pset6-20170423T095251Z/credit.py
import cs50

def main():
    #ask user for valid input
    while True:
        print("Number:", end="")
        f = cs50.get_float()
        if f>0:
            break
    #calculate the sum of the even digits multiplied by 2
    sum = get_even_sum(f)
    
    #add on the sum of the remaining digits
    sum += get_odd_sum(f)
    
    #count the number of digits
    count = get_digits(f)
    
    #find the starting two digits
    two_digit = get_two_digits(f)
    
    #find the first digit
    first_digit = get_first_digit(f)
    
    #checks the validity of the credit card
    credit_card(sum, count, two_digit, first_digit)

#calculate the sum of the even digits according to the Luhn algorithm
def get_even_sum(n):
    product=0
    even_sum=0
    #multiply every even digit by two and then take a cumulative sum
    while n>0:
        n//=10
        product = (n % 10)*2
        while product > 0:
            even_sum += product % 10
            product//=10
        n//=10
    return even_sum

#calculates the sum of the odd digits in a number
def get_odd_sum(n):
    odd_sum=0
    #take the cumulative sum of every odd digit 
    while n>0:
        odd_sum+=(n %10)
        n//=100
    return odd_sum
    
#calculates the number of digits in an umber
def get_digits(n):
    count=0
    while n!=0:
        n//=10
        count+=1
    return count

#finds the first two digits of a number    
def get_two_digits(n):
    while n>100:
        n//=10
    return n

#finds the first digit of a number
def get_first_digit(n):
    n = get_two_digits(n)
    n //=10
    return n

#checks the validity of the credit card according to credit card specifications
def credit_card(sum, count, two_digit, first_digit):
    if sum % 10==0:
        if (count==13 or count==16) and first_digit==4:
            print("VISA")
        elif count==16 and 51<=two_digit<=55:
            print("MASTERCARD")
        elif count==15 and (two_digit==34 or two_digit==37):
            print("AMEX")
        else:
            print("INVALID")
    else:
        print("INVALID")

if __name__ == "__main__":
    main()
    
        
    
    


    
    

    